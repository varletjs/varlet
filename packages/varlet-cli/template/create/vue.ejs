<template>
  <div class="var-<%- kebabCaseName %>"></div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import './<%- kebabCaseName %>.less'

export default defineComponent({
  name: '<%- namespace + bigCamelizeName %>',
})
</script>
