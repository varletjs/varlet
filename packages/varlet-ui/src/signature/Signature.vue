<template>
  <div ref="root" :class="n()">
    <canvas ref="canvas" :height="canvasHeight" :width="canvasWidth"></canvas>
  </div>
</template>

<script lang="ts">
import { defineComponent, onMounted, ref } from 'vue'
import { call, getRect, getStyle } from '@varlet/shared'
import { onWindowResize, useEventListener } from '@varlet/use'
import { createNamespace } from '../utils/components'
import { props } from './props'

const { name, n } = createNamespace('signature')

export default defineComponent({
  name,
  props,
  setup(props) {
    const root = ref<HTMLElement>()
    const canvas = ref<HTMLCanvasElement>()
    const canvasWidth = ref(0)
    const canvasHeight = ref(0)
    let isSigning = false
    let ctx: CanvasRenderingContext2D | null = null

    useEventListener(canvas, 'touchstart', handleTouchstart)
    useEventListener(canvas, 'touchmove', handleTouchmove)
    useEventListener(canvas, 'touchend', handleTouchend)

    onWindowResize(resize)
    onMounted(resize)

    function resize() {
      if (!canvas.value || !root.value) {
        return
      }

      ctx = canvas.value.getContext('2d')
      canvasWidth.value = root.value.offsetWidth
      canvasHeight.value = root.value.offsetHeight
    }

    function handleTouchstart(event: TouchEvent) {
      if (!ctx || !root.value) {
        return
      }

      event.preventDefault()
      isSigning = true

      ctx.beginPath()
      ctx.lineWidth = props.lineWidth
      ctx.strokeStyle = props.strokeStyle === 'currentColor' ? getStyle(root.value).color : props.strokeStyle

      call(props.onStart)
    }

    function handleTouchmove(event: TouchEvent) {
      event.preventDefault()
      if (!canvas.value || !ctx || !isSigning) {
        return
      }

      const clientX = event.touches[0].clientX
      const clientY = event.touches[0].clientY
      const rect = getRect(canvas.value)
      const x = clientX - rect.left
      const y = clientY - rect.top

      ctx.lineTo(x, y)
      ctx.stroke()

      call(props.onSigning, { clientX, clientY, x, y })
    }

    function handleTouchend(event: Event) {
      event.preventDefault()
      isSigning = false
      call(props.onEnd)
    }

    function getDataUrl(canvas: HTMLCanvasElement) {
      switch (props.dataUrlType) {
        case 'png':
          return canvas.toDataURL('image/png')
        case 'jpg':
          return canvas.toDataURL('image/jpeg')
        default:
          return canvas.toDataURL('image/png')
      }
    }

    function isCanvasEmpty(canvas: HTMLCanvasElement) {
      if (!canvas) {
        return true
      }

      const emptyCanvas = document.createElement('canvas')
      emptyCanvas.width = canvas.width
      emptyCanvas.height = canvas.height

      return canvas.toDataURL() === emptyCanvas.toDataURL()
    }

    // expose
    function reset() {
      if (!ctx) {
        return
      }

      isSigning = false
      ctx.clearRect(0, 0, canvasWidth.value, canvasHeight.value)
      ctx.closePath()
    }

    // expose
    function confirm() {
      if (!canvas.value) {
        return
      }

      if (isCanvasEmpty(canvas.value)) {
        return ''
      }

      return getDataUrl(canvas.value)
    }

    return {
      root,
      canvas,
      canvasWidth,
      canvasHeight,
      n,
      confirm,
      reset,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import './signature';
</style>
