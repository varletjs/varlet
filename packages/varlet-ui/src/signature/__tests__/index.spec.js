import { createApp } from 'vue'
import { mount } from '@vue/test-utils'
import { beforeAll, describe, expect, test, vi } from 'vitest'
import Signature from '..'
import VarSignature from '../Signature.vue'

// 模拟 Canvas API
beforeAll(() => {
  // 模拟 getContext
  HTMLCanvasElement.prototype.getContext = vi.fn(() => ({
    clearRect: vi.fn(),
    closePath: vi.fn(),
    beginPath: vi.fn(),
    moveTo: vi.fn(),
    lineTo: vi.fn(),
    stroke: vi.fn(),
  }))

  // 模拟 toDataURL
  HTMLCanvasElement.prototype.toDataURL = vi.fn((type) => {
    return type === 'image/jpeg' ? 'data:image/jpeg;base64,test' : 'data:image/png;base64,test'
  })
})

test('signature plugin', () => {
  const app = createApp({}).use(Signature)
  expect(app.component(Signature.name)).toBeTruthy()
})

describe('test signature component props', () => {
  test('signature lineWidth', async () => {
    const wrapper = mount(VarSignature, {
      props: {
        lineWidth: 2,
      },
    })

    expect(wrapper.find('.var-signature')).toBeTruthy()
    await wrapper.setProps({ lineWidth: 4 })
    wrapper.unmount()
  })

  test('signature strokeStyle', async () => {
    const wrapper = mount(VarSignature, {
      props: {
        strokeStyle: '#000',
      },
    })

    expect(wrapper.find('.var-signature')).toBeTruthy()
    await wrapper.setProps({ strokeStyle: '#f00' })
    wrapper.unmount()
  })

  test('signature dataUrlType', async () => {
    const wrapper = mount(VarSignature, {
      props: {
        dataUrlType: 'png',
      },
    })

    expect(wrapper.find('.var-signature')).toBeTruthy()
    await wrapper.setProps({ dataUrlType: 'jpg' })
    wrapper.unmount()
  })
})

describe('test signature component methods', () => {
  test('signature reset method', async () => {
    const wrapper = mount(VarSignature)
    await wrapper.vm.reset()
    wrapper.unmount()
  })

  test('signature confirm method', async () => {
    const wrapper = mount(VarSignature)
    const result = await wrapper.vm.confirm()
    expect(result).toBe('')
    wrapper.unmount()
  })
})
