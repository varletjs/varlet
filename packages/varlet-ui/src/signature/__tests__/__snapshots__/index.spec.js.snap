// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`signature base function 1`] = `"<div class=\\"var-signature\\"></div>"`;
